File: solutions/csharp/perfect-numbers/10/PerfectNumbers.cs
using System;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0) throw new ArgumentOutOfRangeException();
        return (number - GetAliquotSum(number)) switch
        {
            > 0 => Classification.Deficient,
            < 0 => Classification.Abundant,
            _ => Classification.Perfect,
        };
    }

    private static int GetAliquotSum(int number) =>
        Enumerable
        .Range(1, number / 2)
        .Where(factor => number % factor == 0)
        .Sum();
}

File: solutions/haskell/perfect-numbers/4/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification (..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify num
  | num < 1 = Nothing
  | x < num = Just Deficient
  | x > num = Just Abundant
  | otherwise = Just Perfect
  where
    x = sumOfFactors num

sumOfFactors :: Int -> Int
sumOfFactors num = sum (factors num)

factors :: Int -> [Int]
factors number =
  [divisor | divisor <- [1 .. (number `quot` 2)], number `rem` divisor == 0]